blocks/api/factory.js

```diff
@@ -45,9 +45,15 @@ export function switchToBlockType( block, blockType ) {
 		return null;
 	}
 
+	const attributes = transformation.transform( block.attributes );
+	if ( attributes instanceof Error ) {
+		// Blocks can perform validation and cancel transformations if needed.
+		return attributes;
```

Comment:
What would you suggest returning instead of `null`?  To me this makes sense because it is a generic failure case which doesn't have specific meaning to the user (something went wrong in the code).

I'll address the rest of your comments below.